pkg/controller/util/endpoint/controller_utils_test.go

```diff
@@ -224,3 +231,249 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
+					Selector: map[string]string{
+						"app": "test-0",
+					},
+				},
+			},
+			expect: sets.String{"test/service-0": sets.Empty{}, "test/service-1": sets.Empty{}},
+		},
+		{
+			name:   "get servicesMemberships for pod-1 after service-1 updated",
+			pod:    pods[1],
+			expect: sets.String{},
+		},
+		{
+			name:   "get servicesMemberships for pod-2",
+			pod:    pods[2],
+			expect: sets.String{"test/service-2": sets.Empty{}},
+		},
+		{
+			name:   "get servicesMemberships for pod-2 after service-2 deleted",
+			pod:    pods[2],
+			action: Delete,
+			service: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "service-2",
+					Namespace: "test",
+				},
+				Spec: v1.ServiceSpec{
+					Selector: map[string]string{
+						"app": "test-2",
+					},
+				},
+			},
+			expect: sets.String{},
+		},
+		{
+			name:   "get servicesMemberships for pod-2 after service-5 added",
+			pod:    pods[2],
+			action: Update,
+			service: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "service-5",
+					Namespace: "test",
+				},
+				Spec: v1.ServiceSpec{
+					Selector: map[string]string{
+						"app": "test-2",
+					},
+				},
+			},
+			expect: sets.String{"test/service-5": sets.Empty{}},
+		},
+	}
+
+	for _, test := range tests {
+		if test.service != nil {
+			switch test.action {
+			case Update:
+				if err := fakeInformerFactory.Core().V1().Services().Informer().GetStore().Update(test.service); err != nil {
+					t.Errorf("%s fail to %s service %v.", test.name, test.action, test.service)
+				}
+				// service changed, update it in ServiceSelectorCache
+				_ = cache.Update(fmt.Sprintf("%s/%s", test.service.Namespace, test.service.Name), test.service.Spec.Selector)
+			case Delete:
+				if err := fakeInformerFactory.Core().V1().Services().Informer().GetStore().Delete(test.service); err != nil {
+					t.Errorf("%s fail to %s service %v.", test.name, test.action, test.service)
+				}
+				cache.Delete(fmt.Sprintf("%s/%s", test.service.Namespace, test.service.Name))
+			default:
+				t.Errorf("Unknown action %v", test.action)
+			}
+		}
+		if services, err := cache.GetPodServiceMemberships(fakeInformerFactory.Core().V1().Services().Lister(), test.pod); err != nil || !services.Equal(test.expect) {
+			t.Errorf("%s fail with error %v, get %v but expect %v", test.name, err, services, test.expect)
+		}
+	}
+}
+
+func TestServiceSelectorCache(t *testing.T) {
+	cache := NewServiceSelectorCache()
+	tests := []struct {
+		key      string
+		name     string
+		selector map[string]string
+		action   Action
+		exist    bool
+		expect   labels.Selector
+	}{
+		{
+			key:    "test/service-0",
+			name:   "get a nonexistent selector",
+			action: Get,
+			exist:  false,
+		},
+		{
+			key:      "test/service-0",
+			name:     "add service-0 selector",
+			selector: map[string]string{"app": "test-0"},
+			action:   Update,
+			expect:   labels.Set(map[string]string{"app": "test-0"}).AsSelectorPreValidated(),
```

Comment:
I'm not used to test cases being dependent on the previous case. If there's a way to make test cases independent of each other here and above I think that would be ideal. If not, it would be worth commenting to explain that these test cases depend on each other and the sequence in which they run is important.

Reply:
I added comment to explain cases depend on previous cases, because both selector cache and informer will cache changes in each case(seems more fit with actual scene). It can also undo the change after cases end to make cases independent. What about your opinion?

Reply:
I think I'd prefer an approach where each test case was independent and started with a prepopulated ServiceSelectorCache that was unique for each case. That would ensure that a single failure didn't affect each test case after it. Does that seem reasonable/doable?

Reply:
I choose a backup cache which stores original data to make test cases independent. It is used to restore cache when cache changed during tests.